### ISSUE TITLE
Clean  already exist tmp directory when use s3 rest api when initiate multipart upload use s3 rest api.

### ISSUE THREAD

**Issue by jffree** [Alluxio/alluxio#14477]
According to the description in the aws s3 document, the s3 service will only store the latest data (that is, directly overwrite the existing data). The work of checking whether the data already exists should be done by the client. 

> _Amazon S3 is a distributed system. If it receives multiple write requests for the same object simultaneously, it overwrites all but the last object written. Amazon S3 does not provide object locking; if you need this, make sure to build it into your application layer or use versioning instead._
> [PutObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html)

Therefore, many third-party clients do not support the process of first checking and then overwriting (like spark and mingo).

#14203 In this pr, @ZacBlanco already implement overwrite in `createObjectOrUploadPart` func.  

So, do we also need to do some cleanup operations when initializing multipart upload the request? For example, delete the tmp directory and the multipart upload file.

**Comment by jffree** [comment]
@HelloHorizon @ZacBlanco Please take a look.

**Comment by ZhuTopher** [comment]
From my understanding of the [S3 multipart upload process](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html#mpu-process),
> When you send a request to initiate a multipart upload, Amazon S3 returns a response with an upload ID, which is a unique identifier for your multipart upload.

and

> If you upload a new part using the same part number as a previously uploaded part, the previously uploaded part is overwritten.

So if we somehow erroneously return an upload ID which has an existing tmp directory & parts, then we would directly overwrite them. So to ensure that the upload ID returned is indeed unique, we can use the existence of the tmp directory as the check.

---

So to answer your question,
> So, do we also need to do some cleanup operations when initializing multipart upload the request? For example, delete the tmp directory and the multipart upload file.

I think the answer is no. Instead, we would want to return some sort of error/warning message and retry with another upload ID.

@ZacBlanco wdyt?

**Comment by jffree** [comment]
@ZhuTopher @HelloHorizon 
When promote alluxio, there is a very important point that we hope that the user program does not need to make any changes. If a multipart upload failed and user want to retry again, I thought we should do some clean job, but not return `xxx_s3_multipart_tmp already exists ` error. User do not know why this temporary directory exists, and why it will cause their application to fail.  

#14328 This pr is used to save storage space, when multipart upload failure but ignored by user. This timeout period may be set to a few days or even longer.

**Comment by ZacBlanco** [comment]
I agree with @jffree  -- We shouldn't fail a multipart upload for the same path if the tmp directory already exists.

The way I see it, ideally, each multipart upload could get its own directory (via some UUID?[upload id?]), and then only when completing the multipart upload is all of the data overwritten. Users shouldn't need to modify their applications to handle Alluxio's behavior. We should probably change the logic to allow several ongoing multipart uploads.

Does S3 allow multiple concurrent multipart uploads on the same key?

**Comment by jffree** [comment]
@ZacBlanco @ZhuTopher 
The common scenario for this exception `xxx_s3_multipart_tmp already exists` is: the application exits abnormally when use multipart apt to upload a file, and this error occurs when the application restarts and uploads the file again.
If we support several ongoing multipart uploads,  conflicts are prone to occur when complete multipart upload, and not easy to handle.

**Comment by ZhuTopher** [comment]
@ZacBlanco @jffree Let me clarify my suggestion, I think you might have misunderstood.
> I think the answer is no. Instead, we would want to return some sort of error/warning message and retry with another upload ID.

By this I mean when someone uses the Alluxio REST API to initiate an S3 multipart upload, if the generated upload ID we create happens to have an existing `xxx_s3_multipart_tmp` directory then we should log that and then try a different upload ID until we find one that is not part of an existing multipart upload.

The end-user will not be aware of any of these failure messages and will receive an upload ID like normal.

As for the question:
> Does S3 allow multiple concurrent multipart uploads on the same key?

What do you mean by key? If you mean the `<upload ID:part number>` tuple then as I mentioned the AWS docs say:
> If you upload a new part using the same part number as a previously uploaded part, the previously uploaded part is overwritten.

But if you mean just the `<upload ID>` then there should be no issue with having concurrent uploads.

**Comment by jffree** [comment]
@ZhuTopher 
> By this I mean when someone uses the Alluxio REST API to initiate an S3 multipart upload, if the generated upload ID we create happens to have an existing xxx_s3_multipart_tmp directory then we should log that and then try a different upload ID until we find one that is not part of an existing multipart upload.

This will cause a very serious problem: if the upload operation continues to fail for some reason, temporary directories and files will be created continuously and consume a lot of storage resources.

**Comment by ZhuTopher** [comment]
@jffree That's the same behavior as S3 itself. If you have lots of multipart uploads which you do not mark as completed or aborted, then they will continue to consume storage and you will be charged for it. From [the AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html):
> After you initiate a multipart upload and upload one or more parts, you must either complete or stop the multipart upload in order to stop getting charged for storage of the uploaded parts. Only after you either complete or stop a multipart upload will Amazon S3 free up the parts storage and stop charging you for the parts storage.

That's why you created and we merged in https://github.com/Alluxio/alluxio/pull/14328; to automatically reclaim those failed multipart uploads.

**Comment by jffree** [comment]
Please let me rephrase my point.

My understanding is that multipart upload is the same as direct upload, and both should be overwritten directly. That is, the user should be responsible for his uploading behavior. As aws s3 mentioned in the documentation:
> Amazon S3 is a distributed system. If it receives multiple write requests for the same object simultaneously, it overwrites all but the last object written. Amazon S3 does not provide object locking; if you need this, make sure to build it into your application layer or use versioning instead.

About try a different upload ID until we find one that is not part of an existing multipart upload. This may occupy more storage resources due to the continuous upload failure of one file.
> By this I mean when someone uses the Alluxio REST API to initiate an S3 multipart upload, if the generated upload ID we create happens to have an existing xxx_s3_multipart_tmp directory then we should log that and then try a different upload ID until we find one that is not part of an existing multipart upload.

About #14328, it is not suitable for the currently discussed scenario, it is more suitable for the scenario where the user failed to upload the file but did not retry.

Based on the above considerations, I would think that directly cleaning the tmp directory would be the simplest and most effective way.

**Comment by HelloHorizon** [comment]
I think the clean-up of tmp file can be a feasible method according to the AWS S3 standard behavior. I list some questions here based on my understanding:
1. When do you want to do the cleanup for the tmp file? Will we do it after multi-upload failed? Or user try to upload the same file with tmp file existing?
2. If you decide to clean up the tmp file after multi-load failed? What if the clean up request failed?
3. If you decide to clean up with the lazy way till next retry? Is it acceptable if a user left some tmp there till clean up?

Also add @ggezer into discussion

**Comment by jffree** [comment]
@HelloHorizon 

> When do you want to do the cleanup for the tmp file? Will we do it just after multi-upload failed? Or do it when the user tries to upload the same file with tmp file existing?

There are two scenarios here: one is that the user fails to upload the file and does not retry the operation. At this time, the tmp directory should be cleaned by lazy scanning. The other is that the user fails to upload the file and then retry the upload. At this time, the tmp directory should be cleaned up immediately, without affecting the retry operation.

> If you decide to clean up the tmp file after multi-load failed? What if the clean up request failed?

If the multipart upload fails, the tmp directory will not be cleaned up immediately, because it may just be that the upload of one part failed, and the proxy will perform an overwrite operation when re-uploading the part. 
If the operation of cleaning up tmp fails, there should be an unhandled exception (such as a storage system crash). The maintenance personnel should handle it at this time.

> If you decide to clean up with the lazy way till next retry? Is it acceptable if the user left some tmp there till clean up?

Lazy way means that the user ignored the upload failure error (perhaps this kind of file is not important). This should be scanned and processed by Cleaner (#14328 ).

**Comment by ZhuTopher** [comment]
@jffree I recall you had a PR with these changes already, or did I misremember that? Either way I'm in agreement with:
> There are two scenarios here: one is that the user fails to upload the file and does not retry the operation. At this time, the tmp directory should be cleaned by lazy scanning (#14328 ). The other is that the user fails to upload the file and then retry the upload. At this time, the tmp directory should be cleaned up immediately, without affecting the retry operation.

**Comment by jffree** [comment]
@ZhuTopher 
#14487 This pr (not merged) is used to clean up the tmp directory when user fails to upload the file and then retry the upload. 
#14328 This pr (merged) is used to clean up the tmp directory when user fails to upload the file and does not retry the operation.
